### DISCUSSION TITLE
Wasn't sure where to ask for help, no one seemed to be in IRC

### DISCUSSION THREAD

**Discussion by TheNuclearNexus** [CE-Programming/toolchain#338] [Q&A] [answered]
I've been getting an error 
```Custom error: section .bss is 100024 bytes, 39334 bytes larger than the maximum size of 60690 bytes.```
but I'm not sure what is causing it, I'm new to CE-Programming and the C language in general, any help would be greatly appreciated.

**Comment by adriweb** [comment, marked as answer]
Well you have too many globals (that go in the .bss section), apparently.

You should try not to use globals.

See here: https://ce-programming.github.io/toolchain/static/coding-guidelines.html#avoid-global-variables
and there: https://ce-programming.github.io/toolchain/static/faq.html#what-is-the-c-runtime-memory-layout

As you can see, the heap (memory for dynamic allocation, what `malloc` uses) and the bss (space for uninitialized static data, both variables and constants) share the same memory space, and is limited to around 61 KB.
If you go above that, well, you get this error message.

**Reply by TheNuclearNexus** [reply]
I currently have my array of tiles as a global
```c
#define worldSizeX 100
#define worldSizeY 100
typedef struct {
    int16_t x;
    int16_t y;
    int8_t dist;
    enum Type type;
} Tile;

static Tile tiles[worldSizeX * worldSizeY];
```

**Reply by TheNuclearNexus** [reply]
I'm not sure how I would do it without a global

**Reply by adriweb** [reply]
Are you trying to reinvent tilemaps? Because the toolchain provides the feature (and more) via the graphx library. Check out the corresponding examples: https://github.com/CE-Programming/toolchain/tree/master/examples/graphx

As for big data (sprite resources, tiles...) they can generally live as constant data stored in appvars, separate from the actual program.

**Reply by commandz0** [reply]
Each tile is 8 bytes total. 100x100x8 is 80000, which is about 20 KB more than fit in BSS. You'll probably want to either reduce the size of your world, or split it into chunks that you can store in appvars.

**Reply by TheNuclearNexus** [reply]
the tiles are rendered as isometric squares so I don't believe tilemap would help, as well the data can't be hardcoded since its procedurally generated via perlin noise

**Reply by runer112** [reply]
Why does each tile explicitly store its x and y value? Would those not be implicit in its index in the tilemap? If you could remove those, the struct would be much smaller.

Also, I'm guessing that the enum is being stored as a 3-byte value, while I bet you could get away with a single byte. You can make it a single byte by making it a bitfield with `enum Type type : 8;`, or just by changing its type to `int8_t`.

**Reply by TheNuclearNexus** [reply]
you're right, I could just calculate the x,y in the array from the actual index

**Reply by TheNuclearNexus** [reply]
didn't think about that
